core/templates/pages/exploration-player-page/services/conversation-flow.service.spec.ts

```diff
@@ -0,0 +1,103 @@
+import {NO_ERRORS_SCHEMA} from '@angular/core';
+import {TestBed, waitForAsync} from '@angular/core/testing';
+
+import {ConversationFlowService} from './conversation-flow.service';
+import {StateCard} from 'domain/state_card/state-card.model';
+import {ContentTranslationLanguageService} from './content-translation-language.service';
+import {ContentTranslationManagerService} from './content-translation-manager.service';
+import {ExplorationPlayerStateService} from './exploration-player-state.service';
+import {PlayerTranscriptService} from './player-transcript.service';
+import {TranslateService} from '@ngx-translate/core';
+import {MockTranslateService} from 'components/forms/schema-based-editors/integration-tests/schema-based-editors.integration.spec';
+import {Interaction} from 'domain/exploration/InteractionObjectFactory';
+
+describe('Conversation flow service', () => {
+  let contentTranslationLanguageService: ContentTranslationLanguageService;
+  let contentTranslationManagerService: ContentTranslationManagerService;
+  let conversationFlowService: ConversationFlowService;
+  let explorationPlayerStateService: ExplorationPlayerStateService;
+  let playerTranscriptService: PlayerTranscriptService;
+
+  let displayedCard = new StateCard(
+    null,
+    null,
+    null,
+    new Interaction([], [], null, null, [], '', null),
+    [],
+    null,
+    '',
+    null
+  );
+
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      imports: [HttpClientTestingModule],
+      providers: [
+        ConversationFlowService,
+        {
+          provide: TranslateService,
+          useClass: MockTranslateService,
+        },
+      ],
+      schemas: [NO_ERRORS_SCHEMA],
+    });
+
+    contentTranslationLanguageService = TestBed.inject(
+      ContentTranslationLanguageService
+    );
+    contentTranslationManagerService = TestBed.inject(
+      ContentTranslationManagerService
+    );
+    conversationFlowService = TestBed.inject(ConversationFlowService);
+    explorationPlayerStateService = TestBed.inject(
+      ExplorationPlayerStateService
+    );
+    playerTranscriptService = TestBed.inject(PlayerTranscriptService);
+  }));
+
+  it('should handle new card addition', () => {
+    spyOn(playerTranscriptService, 'addNewCard');
+    spyOn(explorationPlayerStateService, 'getLanguageCode').and.returnValue(
+      'en'
+    );
+    spyOn(
+      contentTranslationLanguageService,
+      'getCurrentContentLanguageCode'
+    ).and.returnValue('es');
+    spyOn(
+      contentTranslationManagerService,
+      'displayTranslations'
+    ).and.returnValue();
+
+    conversationFlowService.addNewCardAndDisplayTranslations(displayedCard);
+    expect(playerTranscriptService.addNewCard).toHaveBeenCalled();
+    expect(
+      contentTranslationManagerService.displayTranslations
+    ).toHaveBeenCalled();
+  });
+
+  it('should tell if supplemental card is non empty', () => {
+    expect(
```

Comment:
Test different scenarios properly. Create a card that has a supplemental interaction and validate that the result is true, create one that has a non-supplemental interaction and validate that the result is false. Try to have your tests be comprehensive in testing behaviours.

Reply:
@Vir-8 It looks like you're using an invalid interaction ID (empty string) to test the false case.

I suggest that you have a private createCard() function in this test that takes the interaction ID. Then call that with null, 'ImageClickInput' and "TextInput' and verify that the correct thing is output each time.

In general, when you test something, please think through the specific behaviours that need to be tested and cover those behaviours properly, otherwise a lot of reviews will go back and forth on things like this. Thanks!

Reply:
Done